### PR TITLE
bpo-14916: use specified tokenizer fd for file input

### PR DESCRIPTION
**Pull request by pmp-p**
PyRun_InteractiveOne*() functions allow to explicitily set fd instead of stdin.
but stdin was hardcoded in readline call.

This patch does not fix target file for prompt unlike original bpo one : prompt fd is unrelated to tokenizer source which could be read only.

Also actual  documentation say "prompt the user" so one would expect prompt to go on stdout not a file for both  PyRun_InteractiveOne*()  and PyRun_InteractiveLoop*().

EDIT/ to be clear this is a &#x1F534;type-bugfix&#x1F534;

https://bugs.python.org/issue14916

### CODE REVIEW

**Comment by pablogsal** [comment]
Hi, @pmp-p for the contribution. could you please add some test that exercises this?

**Comment by pablogsal** [comment]
Also, please add a NEWS entry describing the fix

**Comment by ta0kira** [comment]
I think that "interactive" implies that `fp` should be R/W, or the prompt writes should just be allowed to fail internally if the file is opened R/O.

I'm still not sure if the blocking issue exists (see https://bugs.python.org/issue14916#msg184410), but it seems like the interactive execution should really use separate input and output descriptors, with a target use-case being a PTY descriptor that's been duplicated. Since that might require a lot of changes, I think a sufficient compromise would be to use `fp` for both input and output.

(BTW, I have no say in the approval of this PR; I'm just the one who originally created https://bugs.python.org/issue14916.)

**Comment by pmp-p** [comment]
@pablogsal , here's the test i use :
```
#!/bin/sh

PYVER=${PYVER:-3.8}

echo "testing bpo-14916 for python${PYVER}"


cat <<END > test-bpo-14916.c
#include <stdio.h>
#include <stdlib.h>

#define PY_SSIZE_T_CLEAN
#include <Python.h>

#define MAX 1024
char buf[MAX];

#define TEST_CODE "file_io"

int
main(int argc, char *argv[])
{
    wchar_t *program = Py_DecodeLocale(argv[0], NULL);
    if (program == NULL) {
        fprintf(stderr, "Fatal error: cannot decode argv[0]\n");
        exit(1);
    }
    Py_SetProgramName(program);  /* optional but recommended */
    Py_Initialize();

    FILE *one_file = fopen(TEST_CODE,"w+");

    setenv("BPO14916","FOUND!",1);

    fputs("__import__('os').environ['BPO14916']='not present'\n", one_file);

    fflush(one_file);
    fclose(one_file);



    one_file = fopen(TEST_CODE, "r");
    int line =  0;

    if (one_file) {
        while( fgets(&buf[0], MAX, one_file) ) {
            fprintf( stderr, "%d: %s",++line, buf );
        }


        rewind(one_file);
        int res = PyRun_InteractiveOne( one_file, TEST_CODE);

       fclose(one_file);
    } else {
        fprintf( stderr, "IO error\n");
    }


    printf("\nStatus of bug 14916 : %s\n\n", getenv("BPO14916") );

    if (Py_FinalizeEx() < 0) {
        exit(120);
    }
    PyMem_RawFree(program);
    return 0;
}

END

gcc -o test-bpo-14916 test-bpo-14916.c -I/usr/local/include/python${PYVER} -L/usr/local/lib -lpython${PYVER} -lpthread -ldl -lm -lutil

echo i_should_not_be_parsed | ./test-bpo-14916
```

**Comment by pmp-p** [comment]
attaching file because it seems the bpo parser modify code blocks
[test.txt](https://github.com/python/cpython/files/5205307/test.txt)

**Comment by pmp-p** [comment]
@pablogsal sorry to bother pinging you, but that bug is getting annoying for a repl based testuite i'd like to demonstrate for zipimport, android and web. Do you need another kind of test instead of C one ?

**Comment by pmp-p** [comment]
As @JulienPalard pointed it out, I don't think there should be an implicit fallback to stdin but it was in original PR. It would be nice to have any reviewer's advice.

File: Parser/tokenizer.c [hunk 1]
```diff
@@ -849,7 +849,7 @@ tok_nextc(struct tok_state *tok)
             return Py_CHARMASK(*tok->cur++);
         }
         if (tok->prompt != NULL) {
-            char *newtok = PyOS_Readline(stdin, stdout, tok->prompt);
+            char *newtok = PyOS_Readline(tok->fp? tok->fp : stdin, stdout, tok->prompt);
```
Comment (JulienPalard) [inline Parser/tokenizer.c]:
```suggestion
            char *newtok = PyOS_Readline(tok->fp ? tok->fp : stdin, stdout, tok->prompt);
```

**Review by JulienPalard — CHANGES_REQUESTED** [review]
LGTM (but the space).

Tested on master. It make sense to me because the fd is given to PyParser_ASTFromFileObject up to PyTokenizer_FromFile which stores it in `tok->fp`.

Also on line 835 there's a similar condition : read from the file if tok->fp is not null, else read from stdin.

**Comment by pmp-p** [comment]
I have made the requested changes; please review again

**Review by pablogsal — CHANGES_REQUESTED** [review]
This is still missing an automated test. Please, include one in the test suite.

**Comment by pmp-p** [comment]
@pablogsal sorry, i don't know how to test C-API **from cpython testsuite** on official platforms, i'm just a user of it. Also the only time i had to spend was on the  automated C linux above.

**Comment by pablogsal** [comment]
> @pablogsal sorry, i don't know how to test C-API **from cpython testsuite** on official platforms, i'm just a embedding user of it. Also the only time i had to spend was on the automated C linux above.

You need to add your test in C in this module and expose it to Python (check other functions in the same file):

https://github.com/python/cpython/blob/master/Modules/_testcapimodule.c

And then you can test it here, for instance:

https://github.com/python/cpython/blob/master/Lib/test/test_capi.py

**Comment by pmp-p** [comment]
@pablogsal Thanks a lot for the pointers, but sadly i don't know if and when i'll have time to dig around the C-API as in it's current state after a quick glance it looks very complicated and not very tied (nor appealing)  to my current use of CPython.

 On a side note I think that PR is still missing the type-bugfix label because the very high level api is documented and available to link and call by user programs.

**Comment by pablogsal** [comment]
> state after a quick glance it looks very complicated and not very tied (nor appealing) to my current use of CPython.
> 
> On a side note I think that PR is still missing the type-bugfix label because the very high level api is documented and available to link and call by user programs.

Is fine, don't worry. I can try to see if I can add the test my self and merge the PR. Thanks a lot for the patch and the effort :)

**Comment by pmp-p** [comment]
if it's fixed it 3.10 feel free to close

**Comment by pmp-p** [comment]
i just tried 3.11.0a4 and the bug is still present, it could be quite annoying for https://github.com/ethanhs/python-wasm to feed standard input in a non-blocking way.

File: Misc/NEWS.d/next/C API/2020-09-11-02-50-41.bpo-14916.QN1Y03.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+fixed bug in "The Very High Level Layer" of c-api that prevented
+PyRun_InteractiveOne*() parsing from the provided FD.
```
Comment (arhadthedev) [inline Misc/NEWS.d/next/C API/2020-09-11-02-50-41.bpo-14916.QN1Y03.rst]:
```suggestion
PyRun_InteractiveOne\*() parsing from the provided FD.
```

GitHub already tries to interpret a star as a directive lasting until the end of a line. To be safe with pydoc generator I think it's better to escape it.

**Comment by pablogsal** [comment]
I think the rebase has gone wrong because the diff is too big.

Could you please fix it? I can review and land it afterwards :)

(And sorry for the wait, we have been very busy recently :( )

**Pull request closed without merging**